core/lib/zksync_core/src/consensus/storage/mod.rs

```diff
@@ -0,0 +1,424 @@
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl PersistentBlockStore for BlockStore {
+    async fn state(&self, ctx: &ctx::Ctx) -> ctx::Result<BlockStoreState> {
+        // Ensure that genesis block has consensus field set in postgres.
+        let mut storage = CtxStorage::access(ctx, &self.inner.pool)
+            .await
+            .wrap("access()")?;
+        let first = storage
+            .first_certificate(ctx)
+            .await
+            .wrap("first_certificate()")?
+            .context("store is empty")?;
+        let last = storage
+            .last_certificate(ctx)
+            .await
+            .wrap("last_certificate()")?
+            .context("store is empty")?;
+        Ok(BlockStoreState { first, last })
+    }
+
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<validator::FinalBlock> {
+        let storage = &mut CtxStorage::access(ctx, &self.inner.pool)
+            .await
+            .wrap("access()")?;
+        let justification = storage
+            .certificate(ctx, number)
+            .await
+            .wrap("certificate()")?
+            .context("not found")?;
+        let payload = storage
+            .payload(ctx, number, self.inner.operator_address)
+            .await
+            .wrap("payload()")?
+            .context("miniblock disappeared from storage")?;
+        Ok(validator::FinalBlock {
+            payload: payload.encode(),
+            justification,
+        })
+    }
+
+    /// If actions queue is set (and the block has not been stored yet),
+    /// the block will be translated into a sequence of actions.
+    /// The received actions should be fed
+    /// to `ExternalIO`, so that `StateKeeper` will store the corresponding miniblock in the db.
+    ///
+    /// `store_next_block()` call will wait synchronously for the miniblock.
+    /// Once miniblock is observed in storage, `store_next_block()` will store a cert for this
+    /// miniblock.
+    async fn store_next_block(
+        &self,
+        ctx: &ctx::Ctx,
+        block: &validator::FinalBlock,
+    ) -> ctx::Result<()> {
+        // This mutex prevents concurrent store_next_block calls.
+        let mut cursor = ctx.wait(self.cursor.lock()).await?;
+        if let Some(cursor) = &mut *cursor {
+            let number = MiniblockNumber(
+                u32::try_from(block.header().number.0)
+                    .context("Integer overflow converting block number")?,
+            );
+            let payload =
+                Payload::decode(&block.payload).context("Failed deserializing block payload")?;
+            let want = cursor.inner.next_miniblock;
+            if want < number {
+                return Err(anyhow::anyhow!("expected {want:?}, got {number:?}").into());
+            }
+            if want == number {
+                let block = FetchedBlock {
```

Comment:
Nit: IMO, it'd still make sense to extract the `Payload` -> `FinalBlock` conversion as in the previous implementation so that it's more maintainable.